src/crimson/osd/osd.cc

```diff
@@ -734,6 +734,26 @@ seastar::future<bufferlist> OSD::load_map_bl(epoch_t e)
   }
 }
 
+seastar::future<std::map<epoch_t, bufferlist>> OSD::load_map_bls(
+  epoch_t first,
+  epoch_t last)
+{
+  std::map<epoch_t, bufferlist> bls;
+  epoch_t i = last;
+  for (; i >= first; i--) {
+    if (std::optional<bufferlist> found = map_bl_cache.find(i); found) {
+      bls.emplace(i, *found);
+    } else {
+      break;
+    }
+  }
+  return meta_coll->load_maps(first, i).then(
+    [bls=std::move(bls)](auto&& loaded_maps) mutable {
+    bls.merge(loaded_maps);
+    return std::move(bls);
+  });
+}
```

Comment:
probably we could use `map_reduce()` here?
```c++
seastar::future<std::map<epoch_t, bufferlist>> OSD::load_map_bls(epoch_t first,
                                                                 epoch_t last)
{
  return seastar::map_reduce(boost::make_counting_iterator<epoch_t>(first),
                             boost::make_counting_iterator<epoch_t>(last + 1),
                             [this](epoch_t e) {
    if (auto found = map_bl_cache.find(e); found) {
      return seastar::make_ready_future<pair<epoch_t, bufferlist>>(
        std::make_pair(e, std::move(*found)));
    } else {
      return load_map_bl(e).then([e](auto&& bl) {
        return seastar::make_ready_future<pair<epoch_t, bufferlist>>(
          std::make_pair(e, std::move(bl)));
      });
    }
  },
  std::map<epoch_t, bufferlist>{},
  [](auto&& bls, auto&& epoch_bl) {
    bls.emplace(std::move(epoch_bl));
    return std::move(bls);
  });
}
```

Reply:
I think this may involve some operations done twice, like map_bl_cache.find, right?

Reply:
@xxhdx1985126 could you be more specific? why `map_bl_cache.find()` could be called twice with the same epoch?

Reply:
@tchaikov there's another map_bl_cache.find() in OSD::load_map_bl()

Reply:
but i think we can skip the map_bl_cache.find() in load_map_bls